### PR TITLE
update examples

### PR DESCRIPTION
**Pull request by kureta**
Using `ArgumentParser` and `hparams` as defined in the [Hyperparameters](https://pytorch-lightning.readthedocs.io/en/stable/hyperparameters.html) section of
the documentation. This way we can set trainer flags (such as `precision`,
and `gpus`) from the command line. This also enables tensorboard to log hyperparameters for each experiment.

### CODE REVIEW

File: pl_examples/domain_templates/generative_adversarial_net.py [hunk 1]
```diff
@@ -8,12 +8,11 @@
 """
 import os
 from argparse import ArgumentParser, Namespace
-from collections import OrderedDict
 
 import numpy as np
 import torch
 import torch.nn as nn
-import torch.nn.functional as F
+import torch.nn.functional as F  # noqa
```
Comment (Borda) [inline pl_examples/domain_templates/generative_adversarial_net.py:15]: is it used somewhere?

**Review by Borda — APPROVED** [review]
lgtm

File: pl_examples/domain_templates/generative_adversarial_net.py (hunk 1 above)
Comment (kureta) [reply pl_examples/domain_templates/generative_adversarial_net.py:15]: If you mean `F` yes. In the loss function. If you mean `#noqa` it is to stop my linter from bothering me :)

https://github.com/kureta/pytorch-lightning/blob/24ec0704129032ddf3090fe6904815ca4ea0f47f/pl_examples/domain_templates/generative_adversarial_net.py#L107

**Comment by rohitgr7** [comment]
@kureta can you split the data part in `LightningDataModule`?

**Comment by kureta** [comment]
@rohitgr7 sure

**Comment by kureta** [comment]
I am working on it but I have seen that the `LightningDataModule` has `add_argparse_args` and `from_argparse_args` class methods, just like the `Trainer` class, and they inspect the `__init__` signature and generate arguments automatically. This makes the class' init signature much more useful for the user.

```
    @staticmethod
    def add_model_specific_args(parent_parser: ArgumentParser):
        parser = ArgumentParser(parents=[parent_parser], add_help=False)
        parser.add_argument("--lr", type=float, default=0.0002, help="adam: learning rate")
        parser.add_argument("--b1", type=float, default=0.5,
                            help="adam: decay of first order momentum of gradient")
        parser.add_argument("--b2", type=float, default=0.999,
                            help="adam: decay of first order momentum of gradient")
        parser.add_argument("--latent_dim", type=int, default=100,
                            help="dimensionality of the latent space")

        return parser

    def __init__(self, hparams: Namespace):
        super().__init__()

        self.hparams = hparams
```

Here, the user has to mentally parse `add_argument` methods to see what actual parameters the class takes for init. And throughout the class we refer to all hyperparameters this way: `self.hparams.lr` Whereas in our `LightningDataModule` it is like this:

```
class MNISTDataModule(LightningDataModule):
    def __init__(self, batch_size: int = 64, data_path: str = os.getcwd(), num_workers: int = 4):
        super().__init__()
        self.batch_size = batch_size
        self.data_path = data_path
        self.num_workers = num_workers

        self.transform = transforms.Compose([transforms.ToTensor(),
                                             transforms.Normalize([0.5], [0.5])])
        self.dims = (1, 28, 28)
```

This is super clean. You can see, at a glance, all the initialization arguments. `add_argparse_args` and `from_argparse_args` take care of everything. I think we should have the same functionality for the `LightningModule`.

For now, I'll just rename `add_model_specific_args` to `add_argparse_args` in this file. But I really thing we should implement this for usability and consistency between interfaces.

**Review by tchaton — APPROVED** [review]
Great PR ! Thanks for your contributions to Pytorch Lightning !

**Comment by kureta** [comment]
@tchaton Thanks! Happy to help.

**Review by SkafteNicki — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pl_examples/domain_templates/generative_adversarial_net.py [hunk 2]
```diff
@@ -178,36 +165,63 @@ def on_epoch_end(self):
...
-    parser.add_argument("--b2", type=float, default=0.999,
-                        help="adam: decay of first order momentum of gradient")
-    parser.add_argument("--latent_dim", type=int, default=100,
-                        help="dimensionality of the latent space")
-
-    hparams = parser.parse_args()
-
-    main(hparams)
+
+    # Add pogram level args, if any.
```
Comment (ydcjeff) [inline pl_examples/domain_templates/generative_adversarial_net.py:216]:
```suggestion
    # Add program level args, if any.
```

**Review by ydcjeff — APPROVED** [review]
